### PR TITLE
Add pedestrians and other vehicles

### LINKED ISSUE #618: Add auto-piloted pedestrian and vehicles to the simulation.

### PR DESCRIPTION
**Pull request by emailweixu**
resolve #618

### CODE REVIEW

File: alf/environments/suite_carla.py [hunk 1]
```diff
@@ -1291,10 +1291,30 @@ def __init__(self,
...
         """
         Args:
+            batch_size (int): the number of learning vehicles.
+            map_name (str): the name of the map (e.g. "Town01")
+            vehicle_filter (str): the filter for getting vehicle blueprints.
+            walker_filter (str): the filter for getting walker blueprints.
+            num_other_vehicles (int): the number of autopilot vehicles
+            num_walkers (int): the number of walkers
+            global_distance_to_leading_vehicle (str): the autopiloted vehicles
+                will try to keep such distance from other vehicle.s
```
Comment (le-horizon) [inline alf/environments/suite_carla.py]: vehicle.s => vehicles.

File: alf/environments/suite_carla.py [hunk 2]
```diff
@@ -1291,10 +1291,30 @@ def __init__(self,
...
+            percentage_pedestrians_running (float): percent of running walkers
+            percentage_pedestrians_crossing (float): percent of walkers walking
+                through the road
```
Comment (le-horizon) [inline alf/environments/suite_carla.py]: through => across?

File: alf/environments/suite_carla.py [hunk 3]
```diff
@@ -674,9 +674,9 @@ def __init__(self, world: carla.World):
         self._world = world
         self._map = world.get_map()
         self._waypoints = self._map.generate_waypoints(2.0)
-        self._vehicles = []
-        self._vehicle_dict = {}
-        self._vehicle_locations = {}
+        self._actors = []
```
Comment (le-horizon) [inline alf/environments/suite_carla.py]: comment that actors include both vehicles and pedestrians?

File: alf/environments/suite_carla.py [hunk 4]
```diff
@@ -1379,25 +1420,123 @@ def _spawn(self):
...
+                                                  spawn_points[i].location)
+
+        assert len(self._players) + len(
+            self._other_vehicles) == num_vehicles, (
+                "Fail to create %s vehicles" % num_vehicles)
+
+    def _spawn_walkers(self):
+        walker_blueprints = self._world.get_blueprint_library().filter(
+            self._walker_filter)
+        commands = []
```
Comment (le-horizon) [inline alf/environments/suite_carla.py]: remove unused line: commands = []

**Review by le-horizon — COMMENTED** [review]
Almost good to go.  Some nits.

Thanks,
Le

File: alf/environments/suite_carla.py (hunk 1 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: fixed

File: alf/environments/suite_carla.py (hunk 2 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: fixed

File: alf/environments/suite_carla.py (hunk 3 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: commented

File: alf/environments/suite_carla.py (hunk 4 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: removed

File: alf/environments/suite_carla.py [hunk 5]
```diff
@@ -1359,17 +1400,17 @@ def _spawn(self):
 
         spawn_points = self._world.get_map().get_spawn_points()
         number_of_spawn_points = len(spawn_points)
-        if self._batch_size < number_of_spawn_points:
+
+        num_vehicles = self._batch_size + self._num_other_vehicles
+        if num_vehicles <= number_of_spawn_points:
             random.shuffle(spawn_points)
         elif self._batch_size > number_of_spawn_points:
```
Comment (hnyu) [inline alf/environments/suite_carla.py]: What if num_vehicles > number_of_spawn_points? should we also report errors?

File: alf/environments/suite_carla.py [hunk 6]
```diff
@@ -1379,25 +1420,122 @@ def _spawn(self):
...
+
+    def _spawn_walkers(self):
+        walker_blueprints = self._world.get_blueprint_library().filter(
+            self._walker_filter)
+
+        # 1. take all the random locations to spawn
+        spawn_points = []
+        for i in range(self._num_walkers):
+            spawn_point = carla.Transform()
+            loc = self._world.get_random_location_from_navigation()
```
Comment (hnyu) [inline alf/environments/suite_carla.py:1464]: For a walker with zero speed, will it possibly spawn it in the middle of the road?

**Review by le-horizon — DISMISSED** [review]
LG.

File: alf/environments/suite_carla.py [hunk 7]
```diff
@@ -1379,25 +1420,122 @@ def _spawn(self):
...
+        for response, walker_speed, spawn_point in zip(
+                responses, walker_speeds, spawn_points):
+            if response.error:
+                logging.error(
+                    "%s: %s" % (response.error, spawn_point.location))
+                continue
+            walker = self._world.get_actor(response.actor_id)
+            self._walkers.append({"walker": walker})
+            walker_speeds2.append(walker_speed)
+        walker_speeds = walker_speeds2
```
Comment (hnyu) [inline alf/environments/suite_carla.py:1503]: Why do you need walker_speeds2? It seems to be the same with walker_speeds before this line.

File: alf/environments/suite_carla.py [hunk 8]
```diff
@@ -1379,25 +1420,122 @@ def _spawn(self):
...
+        # 5. initialize each controller and set target to walk to (list is [controler, actor, controller, actor ...])
+        # set how many pedestrians can cross the road
+        self._world.set_pedestrians_cross_factor(
+            self._percentage_walkers_crossing)
+        for walker, walker_speed in zip(self._walkers, walker_speeds):
+            # start walker
+            walker['controller'].start()
+            # set walk to random point
+            location = self._world.get_random_location_from_navigation()
+            walker['controller'].go_to_location(location)
```
Comment (hnyu) [inline alf/environments/suite_carla.py:1533]: I'm curious if after the walker goes to the location, what will it do? Is there any cycle for the walker's movement?

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/environments/suite_carla.py (hunk 5 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: Updated

File: alf/environments/suite_carla.py (hunk 6 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py:1464]: According to the documentation of get_random_location_from_navigation() (https://carla.readthedocs.io/en/0.9.9/python_api/), the location should be a valid position for pedestrian.

File: alf/environments/suite_carla.py (hunk 7 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py:1503]: due to the  possiblity of response.error, walker_speeds2 is a suebsequence of walker_speeds.

File: alf/environments/suite_carla.py (hunk 8 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py:1533]: It's controlled by the ai controller, doing some random walk.

**Review by hnyu — APPROVED** [review]

**Pull request merged**